lambdas/service/app.py

```diff
@@ -688,3 +693,68 @@ def _file_url(uuid, **kwargs):
     return (config.service_endpoint() + '/dss/files' +
             '/' + urllib.parse.quote(uuid, safe='') +
             '?' + urllib.parse.urlencode(kwargs))
+
+
+@app.route('/url', methods=['POST'], cors=True)
+def shorten_query_url():
+    """
+    Take a URL as input and return a (potentially) shortened URL that will redirect to the given URL
+
+    parameters:
+        - name: url
+          in: body
+          type: string
+          description: URL to shorten
+
+    :return: A 200 response with JSON body containing the shortened URL:
+
+    ```
+    {
+        "url": "http://shortened.url/ABC"
+    }
+    ```
+
+    A 400 error is returned if an invalid URL is given.  This could be a URL that is not whitelisted
+    or a string that is not a valid web URL.
+    """
+    try:
+        url = app.current_request.json_body['url']
+    except KeyError:
+        raise BadRequestError('`url` must be given in the request body')
+
+    url_hostname = urllib.parse.urlparse(url).netloc
+    if len(list(filter(lambda whitelisted_url: re.fullmatch(whitelisted_url, url_hostname),
+                       config.url_shortener_whitelist))) == 0:
+        raise BadRequestError('Invalid URL given')
+
+    encoded_url = encode_url(url)
+    storage_service = StorageService(config.s3_public_bucket)
+
+    def get_url_response(path):
+        return {'url': f'http://{config.s3_public_bucket}.s3-website-{aws.region_name}.amazonaws.com/{path}'}
+
+    key_length = 3
+    while key_length <= len(encoded_url):
+        key = f'url/{encoded_url[:key_length]}'
+        try:
+            existing_url = storage_service.get(key)
+        except GetObjectError:
```

Comment:
I agree that `GetObjectError` isn't needed but I do think it's a little cleaner and easier to follow to except `GetObjectError` than `storage_service.client.exceptions.NoSuchKey`.  I don't feel strongly either way though so I can remove it.

Reply:
> follow to except GetObjectError than storage_service.client.exceptions.NoSuchKey

It's shorter but misleading since it creates the impression that we're catching any error that occurred during GetObject. Accuracy over concision, I'd say. If you'd like to keep the custom exception, you could also rename GetObjectError to NoSuchObjectError or similar.